Bump version to v0.16 (#709)
mmtk/mmtk-core: CHANGELOG.md

0.16.0 (2022-12-006)
===

Plan
---

* Refactor `MarkSweep` to work with both our native mark sweep policy and the malloc mark sweep policy backed by malloc libraries.

Allocators
---

* Add `FreeListAllocator` which is implemented as a `MiMalloc` allocator.
* Fix a bug in `ImmixAllocator` that alignment is properly taken into consideration when deciding whether to do overflow allocation.

Policies
---

* Add `MarkSweepSpace`:
  * It uses our native MiMalloc implementation in MMTk core.

>   * When the feature `malloc_mark_sweep` is enabled, it uses the selected mimalloc library to back up its allocation.

Reviewer: It should be "selected malloc library" instead of "selected mimalloc library".

Author: Right. Thanks for pointing this out.

* Malloc mark sweep now accounts for memory based on page usage, and each malloc library may use a different page size.
* The immix space now uses the newly added `BlockPageResource`.
* Changes to Space Function Table (SFT) to improve our boot time and reduce the memory footprint:
  * Rename the current SFT map to `SFTSparseChunkMap`, and only use it for 32 bits architectures.
  * Add `SFTSpaceMap` and by default use it for 64 bits architectures.
  * Add `SFTDenseChunkMap` and use it when we have no control of the virtual address range for a MMTk space on 64 bits architectures.

API
---

* Add an option `thread_affinity` to set processor affinity for MMTk GC threads.
* Add `AllocationSemantics::NonMoving` to allocate objects that are known to be non-moving at allocation time.

> * Add `ReferenceGlue::is_referent_cleared` to allow some bindings to use a special value rathen a normal null reference for a cleared referent.

Reviewer 2: ```suggestion
* Add `ReferenceGlue::is_referent_cleared` to allow some bindings to use a special value rather than a normal null reference for a cleared referent.
```

  * Remove `ObjectModel::OBJECT_REF_OFFSET_BEYOND_CELL`, as we no longer use the raw address of an object reference.
  * Add `ObjectModel::UNIFIED_OBJECT_REFERENCE_ADDRESS`. If a binding uses the same address for `ObjectReference`, `ref_to_address` and `ref_to_object_start`,
    they should set this to `true`. MMTk can utilize this information for optimization.
  * Add `ObjectModel::OBJECT_REF_OFFSET_LOWER_BOUND` to specify the minimam value of the possible offsets between an allocation result and object reference's raw address.
* `destroy_mutator()` no longer requires a boxed mutator as its argument. Instead, a mutable reference to the mutator is required. It is made clear that the binding should
  manage the lifetime of the boxed mutator from a `bind_mutator()` call.
* Remove `VMBinding::LOG_MIN_ALIGNMENT` and `VMBinding::MAX_ALIGNMENT_SHIFT` (so we only keep `VMBinding::MIN_ALIGNMENT` and `VMBinding::MAX_ALIGNMENT`).

Misc
---

* Add a lock-free `BlockPageResource` that can be used for policies that always allocate memory at the granularity of a fixed sized block.

>   The page resource facilitate block allocation and reclamation, and use lock-free operations where possible.

Reviewer 2: ```suggestion
  This page resource facilitates block allocation and reclamation, and uses lock-free operations where possible.
```
